Crypto stack implementation

stze opened the pull request:
This PR introduces a crypto stack implementation based on tweetnacl library.

Commit:
Replace libsodium with tweetnacl to reduce the library footprint

Commit:
Add makekey executable for pub/secr key generation

Public and secret keys are saved in .keys directory

Commit:
Remove libsodium from .travis.yml and enable llvm address sanitizer

Commit:
Remove wrong USE_ prefix from compiler flag

Commit:
This commit introduces the crypto and key agreement protocol implementation.

The crypto module consists of a tunnel packet handler (key agreement) and
crypto_write and crypt_unbox_data function for sending and handling crypted
data.

Commit:
Integrate crypto stack into connection module

* Removed connection_write
* Integrate message_frame handling in parse_cb
* Replace outputstream_write with crypto_write

Commit:
Add test for crypto_tunnel function

Commit:
Makekey now calls crypto_box_keypair instead of doing nothing

Commit:
Replace wrapper function for handle_run and handle_register functional tests

Commit:
Fix undefined behaviour (left shift of negativ value)

Commit:
Merge branch 'master' into 2016-03-stze-crypto-stack-implementation

Commit:
Fixed error in message frame handling and renamed crypto_unbox_data to crypto_read

Commit:
Rename crypto-tunnel.c to crypto.c

Commit:
Add static identifier to crypto variables

Commit:
Fixed error in ciphertext handling

Commit:
Single --wrap functions can now be disabled through a global variable

The global variable(s) allows us to activate a wrapped function for test a and
deactivate it for test b.

Commit:
travis.yml coveralls gardening

Commit:
Add keypair to travis build and let crypto test fail if pub key is not available

Commit:
Merge branch 'master' into 2016-03-stze-crypto-stack-implementation

Commit:
Add docstrings for crypto functions

Commit:
Merge branch '2016-03-stze-crypto-stack-implementation' of github.com:splone/splonebox-core into 2016-03-stze-crypto-stack-implementation

mkind commented on src/sb-makekey.c:
```diff
@@ -0,0 +1,57 @@
+  if (mkdir(".keys",0700) == -1)
+    LOG_ERROR("unable to create .keys directory\n");
+
+  if (filesystem_save_sync(".keys/server-long-term.pub", pk, sizeof(pk))) {
```
Is `pk` already generated here?

stze replied:
no. but this was fixed few commits later

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,275 @@
+int crypto_init(void)
+{
+  if (filesystem_load(".keys/server-long-term", serverlongtermsk,
+      sizeof serverlongtermsk) == -1) {
```
Shouldn't it be `sizeof(serverlongtermsk)`?

stze replied:
that doesn't matter, but feel free to change and add an entry to style guide

mkind replied:
Ah, parentheses are only needed if a type is passed.

mkind commented on CMakeLists.txt:
```diff
@@ -124,6 +124,7 @@ set(SPLONEBOX-SOURCES
   src/rpc/connection/loop.c
```
License preamble is missing.

stze replied:
? where and for what

mkind replied:
Ah.. correct. I meant the file `src/rpc/connection/crypto.c`.

stze replied:
which license preamble is missing?

stze replied:
yeah. please add the preamble.

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,275 @@
+
+int crypto_verify_header(unsigned char *data, uint64_t *length)
+{
+  if (!(byte_isequal(data, 8, "oqQN2kaM")))
```
`"oqQN2kaM"` this conflicts with the [Wiki](https://github.com/splone/splonebox-core/wiki/Crypto).

stze replied:
no, it doesn't. "oqQN2ka**M**" is the message packet identifier and "oqQN2ka**T**" the tunnel packet identifier. wiki is WIP, the message packet is not described yet.

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,275 @@
+unsigned char clientshorttermpk[32];
+
+/* shared secrets */
+unsigned char clientshortserverlong[32];
```
A better name for a variable holding a shared secret would be `sharedsecret`.

stze replied:
i don't agree. there can be more than on shared secret (clientshortservershort, clientshortserverlong), so naming it `sharedsecret` could confuse the reader.

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,275 @@
+  /* init clientshorttermpk */
+  memcpy(clientshorttermpk, data + 24, 32);
+  if (crypto_box_beforenm(clientshortserverlong, clientshorttermpk,
+      serverlongtermsk) != 0)
```
As far as I see, this case can never happen. Although it does not hurt this test might be superfluous.

stze replied:
which case?

mkind replied:
`crypto_box_beforenm` always returns 0. There is now error case.

stze replied:
why does `crypto_box_beforenm` always return 0? please provide the source of the information. if it does alway return 0, please remove the check :wink:

mkind replied:
As far as I see, [it returns the result of crypto_core_hsalsa20](https://github.com/splone/splonebox-core/pull/29/commits/0a798382585bbf557dc8a758808c6b40da89bf8f#diff-554afb1e4833821947cb0813c951ed21R467), which [always returns 0](https://github.com/splone/splonebox-core/pull/29/commits/0a798382585bbf557dc8a758808c6b40da89bf8f#diff-554afb1e4833821947cb0813c951ed21R128).

stze replied:
ok, then please remove the check :+1:

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,275 @@
+    return (-1);
+
+  /* read length */
+  length = uint64_unpack(data + 8);
```
`length` isn't used in this function?

stze replied:
yes. please add a check if `length` is 136, otherwise return -1

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,275 @@
+  uint64_pack(nonce + 16, cc->nonce);
+
+  blocklen = crypto_box_ZEROBYTES + length;
+  block = CALLOC(blocklen, unsigned char);
```
Missing `NULL` check.

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,275 @@
+  blocklen = crypto_box_ZEROBYTES + length;
+  block = CALLOC(blocklen, unsigned char);
+
+  memcpy(block + 32, data, length);
```
It might be a good idea, to replace `32` with `crypto_box_ZEROBYTES`.

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,275 @@
+  uint64_t length;
+
+  /* check if first 8 byte of packet identifier are correct */
+  if (!(byte_isequal(data, 8, "oqQN2kaT")))
```
Identifier bytes like `"oqQN2kaT"` should be macros.

Commit:
Try to effectively zero len bytes starting at pnt even with optimization applied

After use, sensitive data should be overwritten, but memset() and hand-written
code can be silently stripped out by an optimizing compiler or by the linker.

The sbmemzero() function tries to effectively zero len bytes starting at pnt,
even if optimizations are being applied to the code.

Commit:
Gardening crypto.c if statements

mkind committed:
introducing macros for crypto message identifier and nonce prefix

mkind committed:
adding two length checks

mkind committed:
adding missing NULL checks

mkind committed:
allocated mem for tunnel init packet on stack instead of heap

mkind committed:
gardening

mkind commented on test/functional/crypto-tunnel.c:
```diff
@@ -0,0 +1,83 @@
+  crypto_init();
+
+  /* positiv test */
+  assert_int_equal(0, crypto_tunnel(&cc, packet, &write));
```
The spec of the function claims:

> Handle a client tunnel packet and send a server tunnel packet as reponse

Is there a test for the latter one?

stze replied:
yes. see `validate_crypto_tunnel`

mkind committed:
removing unnecessary crypto_box_beforenm checks

mkind committed:
In case of an identifier suffix other than 'T' and 'M' the wrapper
function returns (0) which indicates that everything went fine. Since
only 'T' and 'M' are allowed, it should return an error.

mkind commented:
Before merging, the nonce should be adjusted.

Commit:
Update nonce to an odd/even scheme

* Server uses even nonces
* Client uses odd nonces

stze commented:
If there is no more reason to not merge this into master, please merge. Waiting for the client and synchronzing seems irrelevant or the wrong approach to me, because every developer is able to switch between branches.

bontric commented:
I'm unable to connect two clients at the same time to the core properly. The crypto tunnel message of the second client is considered invalid. 
I get `"sb: establishing crypto tunnel failed"` on the server side. I tested a few combinations and it seems like the first attempt to connect is _always successful_ and every attempt afterwards is _never successful_

The error occurs in `crypto.c:133`

``` c
/* check if box can be opened (authentication) */
if (crypto_box_open_afternm(allzeroboxed, allzeroboxed, 96, nonce, 
    clientshortserverlong))
  goto fail;  
```

The error can be reproduced using this script (modify & clone to test with two connecting plugins):

## Script

``` python
from splonecli import Plugin, RemoteFunction


@RemoteFunction
def foo():
    pass


plug = Plugin(
    "<api_key>",
    "foo",
    "bar",
    "some",
    "thing",
    debug=True)

plug.connect("127.0.0.1", 6666)
plug.register()
plug.listen()
```

Commit:
Remove memzero of public long term key

stze commented:
I've fixed the problem. There was one `memzero`too much :wink:

mkind commented on src/rpc/connection/crypto.c:
```diff
@@ -0,0 +1,323 @@
+  length = uint64_unpack(data + 8);
+  if (length != 136)
+    return (-1);
+
```
A verification of the 64 byte zeros is missing. If the padding is missing or invalid, the server must not respond.

Commit:
Introduce filesystem_open_lock function derived from nacl library

Commit:
Add minutekey and lastminutekey

Commit:
Implement protocol changes to crypto_stack

Commit:
gardening

Commit:
Merge branch 'master' into 2016-03-stze-crypto-stack-implementation

Commit:
Merge branch 'master' into 2016-03-stze-crypto-stack-implementation

mkind committed:
improved log output during parse_cb

Commit:
Initialize lengthpacked with zero

Commit:
Merge branch '2016-03-stze-crypto-stack-implementation' of github.com:splone/splonebox-core into 2016-03-stze-crypto-stack-implementation
